### PR TITLE
[MRG] Add fit_intercept parameter to GLM and GLMCV (see issue #308)

### PR DESCRIPTION
**Pull request by DreHar**
Add fit_intercept parameter to GLM and GLMCV #308

### CODE REVIEW

**Comment by jasmainak** [comment]
can you add a test?

**Comment by jasmainak** [comment]
PR looks reasonable to me although I didn't take a very close look. Let me know once you've added tests.

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -71,20 +71,31 @@ def _probit_g6(z, pdfz, cdfz, thresh=5):
     return res
 
 
-def _lmb(distr, beta0, beta, X, eta):
+def _z(beta0, beta, X, fit_intercept):
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: add a minimal docstring

File: pyglmnet/pyglmnet.py [hunk 2]
```diff
@@ -71,20 +71,31 @@ def _probit_g6(z, pdfz, cdfz, thresh=5):
...
-def _mu(distr, z, eta):
+def _mu(distr, z, eta, fit_intercept):
     """The non-linearity (inverse link)."""
     if distr in ['softplus', 'gamma']:
         mu = np.log1p(np.exp(z))
     elif distr == 'poisson':
         mu = z.copy()
-        intercept = (1 - eta) * np.exp(eta)
-        mu[z > eta] = z[z > eta] * np.exp(eta) + intercept
+        if fit_intercept:
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]:
```
beta0 = (1 - eta) * np.exp(eta) if fit_intercept else 0
```

File: pyglmnet/pyglmnet.py [hunk 3]
```diff
@@ -447,6 +481,10 @@ class GLM(BaseEstimator):
         specifies the scoring metric.
         one of either 'deviance' or 'pseudo_R2'.
         default: 'deviance'
+    fit_intercept : boolean
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: mind the space

File: pyglmnet/pyglmnet.py [hunk 4]
```diff
@@ -819,8 +875,8 @@ def predict(self, X):
             raise ValueError('Input data should be of type ndarray (got %s).'
                              % type(X))
 
-        yhat = _lmb(self.distr, self.beta0_,
-                    self.beta_, X, self.eta)
+        yhat = _lmb(self.distr, self.beta0_, self.beta_, X, self.eta,
+                    fit_intercept=True)
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: why not `fit_intercept=self.fit_intercept`?

Comment (pavanramkumar) [reply pyglmnet/pyglmnet.py]: see line 914

File: pyglmnet/pyglmnet.py [hunk 5]
```diff
@@ -1008,6 +1064,10 @@ class GLMCV(object):
         specifies the scoring metric.
         one of either 'deviance' or 'pseudo_R2'.
         default: 'deviance'
+    fit_intercept : boolean
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: mind the space

File: tests/test_pyglmnet.py [hunk 6]
```diff
@@ -324,6 +325,13 @@ def rmse(a, b):
     y_pg = glm.predict(X)
     assert abs(rmse(Y, y_sk) - rmse(Y, y_pg)) < 1.0
 
+    glm = GLM(distr='gaussian', alpha=l1_ratio, reg_lambda=alpha,
```
Comment (pavanramkumar) [inline tests/test_pyglmnet.py]: what about `max_iter=5`: we're just testing the specific `fit_intercept` behavior right?

File: pyglmnet/pyglmnet.py (hunk 4 above)
Comment (jasmainak) [reply pyglmnet/pyglmnet.py]: that won't work. I suspect line 914 is not properly tested. It needs to be fixed then.

**Comment by jasmainak** [comment]
@pavanramkumar while you're at it don't forget to update `whats_new.rst` giving credit to @DreHar for the original PR

**Comment by jasmainak** [comment]
Thanks @pavanramkumar and @DreHar !

**Pull request merged**